src/librustc_trait_selection/traits/structural_match.rs

```diff
@@ -142,6 +146,22 @@ impl<'a, 'tcx> TypeVisitor<'tcx> for Search<'a, 'tcx> {
                 self.found = Some(NonStructuralMatchTy::Dynamic);
                 return true; // Stop visiting.
             }
+            ty::Foreign(_) => {
+                self.found = Some(NonStructuralMatchTy::Foreign);
+                return true; // Stop visiting
```

Comment:
```suggestion
                return true; // Stop visiting.
```
Please also copy this comment to the other `return true;`. The use of a boolean here is extremely confusing.